src/rules/matchers.ts

```diff
@@ -368,6 +374,68 @@ export class CookieMatcher extends Serializable implements RequestMatcher {
     }
 }
 
+export class CallbackMatcher extends Serializable implements RequestMatcher {
+    readonly type = "callback";
+  
+    constructor(
+        public callback: (request: OngoingRequest) => MaybePromise<boolean>
+        ) {
+        super();
+    }
+  
+    async matches(request: OngoingRequest) {
+        try {
+            return await this.callback(request);
+        } catch (error) {
+            return false;
```

Comment:
This should at least be console logged I think. Or maybe it should actually throw the error, so we fail to handle the request entirely? It's definitely confusing if a matching callback like this unexpectedly stops matching because of an invisible exception.

Reply:
One or the other is fine - if we throw here, it should percolate back up to the normal "request handling failed" here: https://github.com/httptoolkit/mockttp/blob/b769d55a83cc9950e61b32c88188a2bf74db5a81/src/server/mockttp-server.ts#L468.

Ah, hmm, actually I think we might need to pull [this line](https://github.com/httptoolkit/mockttp/blob/b769d55a83cc9950e61b32c88188a2bf74db5a81/src/server/mockttp-server.ts#L438) into that try/catch to make sure that works.

Can you try just throwing here, and adding a test to make sure that you get a 503 with some kind of useful message when that happens?

Reply:
Definitely needs a unit test.

I'm not very familiar with Mocha, the test:

```
    it("should throw a Matcher exception if the callback throws an error", async () => {
        await server.get('/abc').matching(() => {
            throw new Error("Matcher exception");
        }).thenReply(200, 'Mocked response');
        
        const result = await fetch(server.urlFor('/abc'));

        console.log("It should reach here");
        await expect(result).to.have.status(500);
    });
```
Seems to completely end the unit testing framework completely. The exception seems to be unhandled, which is rather odd because if I add logging I can see it's handled as it's saying:

>Failed to handle request: Error: Matcher exception
>    at CallbackMatcher.callback (/home/joe/projects/mockttp/test/integration/matchers/callback-matching.spec.ts:56:19)
>    at CallbackMatcher.<anonymous> (/home/joe/projects/mockttp/src/rules/matchers.ts:391:31)
>    at Generator.next (<anonymous>)
>    at fulfilled (/home/joe/projects/mockttp/dist/rules/matchers.js:5:58)
>    at processTicksAndRejections (node:internal/process/task_queues:96:5)

from the line `console.error("Failed to handle request:", this.debug ? e : e.message); `

It then sets up a 500 response: `response.writeHead(e.statusCode || 500, e.statusMessage || 'Server error');`, and eventually calls the `this.announceResponseAsync(response)`, but then exits with an unhandled exception because of the Matcher exception.

Reply:
Yes, sounds like there must be an unhandled error somewhere, but I have no idea why it'd be half caught like that. There must be some missing handling - obviously until now matchers were never expected to throw errors, so there's no error handling in place there at all.

I'm not sure where that could be coming from just from this description, but feel free to push the broken code here and I can take a closer look if that'd be useful.